src/kernel.rb

```diff
@@ -60,20 +60,31 @@ def warn(*warnings, **)
     nil
   end
 
-  def sprintf(fmt, *positional_args)
-    args = positional_args.dup
+  def sprintf(format_string, *arguments)
+    tokens = []
+
     index = 0
-    format = fmt.chars
-    result = []
-    positional_args_used = nil
-    default_precision = 6
+    chars = format_string.chars
 
-    # Get non-exclusive flags that occur after a leading '%'
-    # but before the width field
-    get_flags = ->(format_chars, index) {
+    current_char = -> { chars[index] }
+    next_char = lambda {
+      index += 1
+      chars[index]
+    }
+    next_number = lambda {
+      c = chars[index]
+      raise ArgumentError, 'bad width' unless ('1'..'9').cover?(c)
+      num = [c]
+      while ('0'..'9').cover?(c = next_char.())
+        num << c
+      end
+      num.join.to_i
```

Comment:
I would make `num` a String instead of an Array, this only requires two changes:
```ruby
num = c
# ...
num.to_i
```
I would argue this is easier to read, and as a bonus it should be more efficient too (we now only use 1 String, instead of 1 Array where each item is a String, so it removes the object creation/allocation from the loop.

Reply:
It should even be possible to remove the String intermediate and just work with an int instead:
```ruby
num = c.to_i
while ('0'..'9').cover?(c = next_char.())
  num = num * 10 + c.to_i
end
num
```
This might be a bit harder to read, and efficiency will now depend on the number of characters we're parsing. I guess in general that would be 1 number, where this should be a little bit faster (it removes a string copy operation). For long numbers this would probably be slower due to the amount of `to_i` calls)